File: week7/identify_poet.pl
#!/usr/bin/perl -w
use strict;

foreach my $arg (@ARGV) {

my $count = 0;
my $words = 0;
my $target = "";
my %poets = ();
my @Tnew;
#print "Start1\n";
	foreach my $poet (glob "poets/*.txt") {
		$poet =~ s/poets\///;
		$poet =~ s/\.txt//;
		$poet =~ tr/_/ /;
		$poets{"$poet"} {scoreisnotaword} = 0;
	}
#print "End1\n";
#print "Start2\n";
	open(T,"<$arg") or die "$0: Can't open $arg: $!\n";
	while (<T>) {
		my @Titems = split(/[^A-Za-z]/, $_);
		foreach (@Titems){
		    if( ( defined $_ ) and !($_ =~ /^$/ )){
				push (@Tnew, $_);
			}
		}
	}
	#print "End2\n";
	#print "Start3\n";
	foreach my $file (glob "poets/*.txt") {
		#print "current poet is: $poet ";
		$words = 0;
		open(F,"<$file") or die "$0: Can't open $file: $!\n";
		my $poet = $file;
		$poet =~ s/poets\///;
		$poet =~ s/\.txt//;
		$poet =~ tr/_/ /;
		my @Fnew;
		while (<F>) {
			my @Fitems = split(/[^A-Za-z]/, $_);
			foreach my $word (@Fitems){
				$word = lc($word);
			    if( ( defined $word ) and !($word =~ /^$/ )){
			        push (@Fnew, $word);
			        $words++;
			        if ( !defined $poets{"$poet"}{"$word"} ) {
			        	$poets{"$poet"}{"$word"} = 1;
			    	} else {
			    		$poets{"$poet"}{"$word"}++;
			    	}
			    }
			}
		}
		$poets{"$poet"}{totalisnotaword} = $words;
		#print "$poet $words\n";
	}
	#print "End3";
	foreach $target (@Tnew) {
		$target = lc($target);
		foreach my $poet (glob "poets/*.txt") {
			$poet =~ s/poets\///;
			$poet =~ s/\.txt//;
			$poet =~ tr/_/ /;
			if ( !defined $poets{"$poet"}{"$target"} ) {
				$poets{"$poet"}{scoreisnotaword} = $poets{"$poet"}{scoreisnotaword} + log( 1 / $poets{"$poet"}{totalisnotaword} );
				#$poets{"$poet"}{"$target"} = $count;
			} else {
				$poets{"$poet"}{scoreisnotaword} = $poets{"$poet"}{scoreisnotaword} + log( ($poets{"$poet"}{"$target"} + 1) / $poets{"$poet"}{totalisnotaword} );
			}
		#print "\n";
		}
	}	
	#foreach my $poet ( sort { $poets{$b}{scoreisnotaword} <=> $poets{$a}{scoreisnotaword} } keys %poets ) {
	#	printf "%s: log_probability of %5.1f for %s\n", $ARGV[0], $poets{"$poet"} {scoreisnotaword}, $poet;
	#}
	my ($poet) = sort { $poets{$b}{scoreisnotaword} <=> $poets{$a}{scoreisnotaword} } keys %poets;

	printf "%s most resembles the work of %s (log-probability=%5.1f)\n", $arg, $poet, $poets{"$poet"}{scoreisnotaword};
}	

File: week7/identify_poet.py
#!/usr/bin/python

import sys
import fileinput
import re
import glob
import string
import operator
import math
import collections

for arg in sys.argv[1:]:

	myhash = collections.defaultdict(dict)
	myhash2 = collections.defaultdict(dict)
	targets = [];
	poets = [];

	for poet in glob.glob("poets/*.txt"):
		m = re.search('(?<=poets/)\w+', poet);
		poet = m.group(0).replace("_", " ")
		myhash2[poet]['scoreisnotaword'] = 0;
		poets.append(poet);


	a = open(arg, 'r');
	inputtext = [];
	for lines in a:
		lines = lines.lower()
		inputtext += re.split(r'[^a-zA-Z]', lines);
	targets += filter(None, inputtext);

	for file in glob.glob("poets/*.txt"):
		words = [];
		input = [];
		poet = file;
		m = re.search('(?<=poets/)\w+', file);
		poet = m.group(0)
		poet = poet.replace("_", " ")
		f = open(file, 'r');
		for lines in f:
			lines = lines.lower()
			input += re.split(r'[^a-zA-Z]', lines);
		words += filter(None, input);
		for k in words:
			myhash[poet][k] = myhash[poet].setdefault(k, 0) + 1
		myhash2[poet]['totalisnotaword'] = len(words)
		#counter = words.count((sys.argv[1]).lower());
		#print "%(counter)4d/%(words)6d = %(frequency).9f %(poet)s" % {"counter": counter, "words": len(words), "frequency": float(counter)/len(words), "poet": poet}
		#print "log((%(counter)d+1)/%(words)6d) = %(frequency)8.4f %(poet)s" % {"counter": counter, "words": len(words),

	d = collections.defaultdict(int)
	for target in targets:
		#print target;
		for poet in glob.glob("poets/*.txt"):
			m = re.search('(?<=poets/)\w+', poet);
			poet = m.group(0).replace("_", " ")
			d[target] = myhash[poet].setdefault(target, 0);
			myhash2[poet]['scoreisnotaword'] += math.log(((float(d[target]) + 1)/myhash2[poet]['totalisnotaword']));

	s = collections.defaultdict(int);
	for poet in poets:	
		s[poet] = myhash2[poet]['scoreisnotaword']

	#for p in sorted(s, key=s.get, reverse=True):
	#	print p

	#for poet in poets:	
		#print myhash2[poet]['scoreisnotaword'];
	#	print "%(poem)s most resembles the work of %(poet)s (log-probability=%(lp)5.1f)\n" % {"poem": arg, "poet": poet, "lp": myhash2[poet]['scoreisnotaword']}

	poet = sorted(s, key=s.get, reverse=True)[0]
	print "%(poem)s most resembles the work of %(poet)s (log-probability=%(lp)5.1f)" % {"poem": arg, "poet": poet, "lp": myhash2[poet]['scoreisnotaword']}


